### ISSUE TITLE
pos2transform in ft_sourceplot fails for iEEG data

### ISSUE THREAD

**Issue by StolkArjen** [fieldtrip/fieldtrip#1385]
Please see below for a couple of lines of code replicating the issue. This code should produce a figure like that shown at step 52 in the human ecog tutorial: http://www.fieldtriptoolbox.org/tutorial/human_ecog/

I could reproduce the error in another dataset, while in yet another dataset it actually produced a figure but the functional data in that figure was projected/interpolated onto the entire mesh rather than limited to the cortical area covering the iEEG electrodes. Glancing at the git history of ft_sourceplot, it's not immediately clear when this issue was introduced. Nor is it clear why pos2transform ought to be called for the iEEG data type in the first place. Per the code below, the anatomical data is given as a secondary input to ft_sourceplot and no transformation matrix is needed? Thoughts?

 
-------------------

datadir = dccnpath('/home/common/matlab/fieldtrip/data/ftp/tutorial/human_ecog');
subj = 'SubjectUCI29';
subj_dir = fullfile(datadir, subj);
cd(subj_dir)

load('SubjectUCI29_freq.mat')
pial_lh = ft_read_headshape('freesurfer/surf/lh.pial');

cfg              = [];
cfg.baseline     = [-.3 -.1];
cfg.baselinetype = 'relchange';
freq_blc = ft_freqbaseline(cfg, freq);

cfg             = [];
cfg.frequency   = [70 150];
cfg.avgoverfreq = 'yes';
cfg.latency     = [0 0.8];
cfg.avgovertime = 'yes';
freq_sel = ft_selectdata(cfg, freq_blc);

cfg              = [];
cfg.funparameter = 'powspctrm';
cfg.funcolorlim  = [-.5 .5];
cfg.method       = 'surface';
cfg.interpmethod = 'sphere_weighteddistance';
cfg.sphereradius = 8;
cfg.camlight     = 'no';
ft_sourceplot(cfg, freq_sel, pial_lh);

the input is freq data with 145 channels, 1 frequencybins and 1 timebins
the call to "ft_selectdata" took 0 seconds and required the additional allocation of an estimated 1 MB
not plotting anatomy
not using an atlas
not using a region-of-interest
The source functional does not contain a triangulated surface, we may need to interpolate to a surface mesh
The logical indices in position 2 contain a true value outside of the array bounds.

Error in pos2transform (line 33)
transform = pos(:,sel)/ind(:,sel);

Error in ft_sourceplot (line 1151)
      functional.transform = pos2transform(functional.pos);

**Comment by schoffelen** [comment]
has this worked before? 
I would expect (without looking at the code in ft_sourceplot) that the variable 'functional' is either a volume or a source datatype, and that based on the internal logic of ft_sourceplot it is determined whether the functional data needs to be interpolated onto the cortical sheet on-the-fly.

**Comment by schoffelen** [comment]
The section of the code that you end up in, indeed assumes the 'functional' data to be 3D-reshapeable (i.e. the values in .pos to be defined on a regular 3D grid), which clearly is not the case in your use-case. Again, I am surprised if this has worked at all before.

**Comment by StolkArjen** [comment]
Yeah, this has worked for, well, at least the past four years or so. But I can't quickly figure out what code change broke it either. Anyway, I was able to reproduce it just now, and it is a crucial step in iEEG data visualization..

Is that transform necessary for projecting electrode activity onto near vertices, or can we add a conditional switch here bypassing its creation?

call: ft_sourceplot(cfg, freq_sel, pial_lh);
<img width="785" alt="Screen Shot 2020-05-12 at 10 29 02 PM" src="https://user-images.githubusercontent.com/3984917/81742249-179c0300-94a0-11ea-9a09-bed16dd8070b.png">

**Comment by schoffelen** [comment]
Rather than just a conditional switch as a band aid, I would like to understand why it worked before, and all of a sudden doesn't anymore. The point is that, apparently, the section 1148-1152 deals with 'isUnstructuredFun', which would mean that the 'pos' in functional are not supposed to be 3D-reshapeable. Then, it does not make sense to try and estimate a transformation matrix from this, because pos2transfrom relies on the assumption that the pos = Nx3, where N is the product of xdim*ydim*zdim of a regular 3D grid. Either, pos2transform has changed such that before it had a safety net, and returned eye(4) in case its regular behavior didn't yield a solution (i.e. the preconditions weren't met), or functional contained a tri before, which caused this section of the code not to be reached (and line 1146-1148 to be executed. I assume that latter.

**Comment by StolkArjen** [comment]
Indeed, pos here pertains to not a grid but arbitrary electrode positions. Your hunch makes sense, yet the only place where the tri field could have been transferred to 'functional' from 'anatomical' is around line 350. Debugging at this stage shows that hasanatomical = 0 and cfg.downsample = 1, meaning no interpolation or downsampling is performed here and, thus, no tri field seems to be ever transferred. Apart from ft_preamble, 'anatomical' is never used in the function. Also, looking at least two years back in the code history, there seem to be no related changes to the handling of 'anatomical' in ft_sourceplot. I'm baffled as to how functional could have ever obtained a tri field in the past.

**Comment by schoffelen** [comment]
Irrespectively, I don't understand the logic of the code:

pos2transform is already called at an earlier instance, only for those input arguments that both have a pos and a dim. This only hold for data that is ~unstructuredfun, which is correct. Now in the lines that cause your issue, the only purpose for that is to determine whether an interpolation to another surface is required, and an attempt to make a transform is made. I think that the second call to pos2transform should go. If anything that line needs to be replaced by transform = eye(4). This will fix the problem for you, I assume, but I don't know whether it will break anything else. Could you make the change (starting from a clean FT copy), check whether the test functions that use ft_sourceplot still work (do a grep on ft_sourceplot for the functions in fieldtrip/test, and run the functions (or use ft_test for that, which I think has functionality to run a selection of test function that have a certain dependency) and if all is fine submit a PR?
Thanks.

**Comment by robertoostenveld** [comment]
I copied the code in a test and can confirm the error. Using
```
  git checkout 'master@{2020-01-01 00:00:00}'
  git checkout 'master@{2019-01-01 00:00:00}'
  git checkout 'master@{2018-01-01 00:00:00}'
  git checkout 'master@{2017-01-01 00:00:00}'
  git checkout 'master@{2016-01-01 00:00:00}'
```
however, I see that the error also applies to older FT versions.

```
The logical indices in position 2 contain a true value outside of the array
bounds.
Error in pos2transform (line 33)
transform = pos(:,sel)/ind(:,sel);
Error in ft_sourceplot (line 1157)
      functional.transform = pos2transform(functional.pos); 
```

So I think we don't have to look at a problem that was introduced in the FT code itself, but rather at the difference between the current use and previous use.

**Comment by StolkArjen** [comment]
Thanks for running this test, and mind boggling results. The code snippets and data originate from the human ECoG tutorial, so way of use hasn't changed for a while either. That is, it's always been the case that one would feed in a freq structure with an elec field, combined with a separate mesh input for the anatomical. Moreover, looking at the if-else statement at line 350 ft_sourceplot, I think the functional and anatomical are supposed to enter than first "if" section (line 341). In fact, forcing this by adding an 'anatomy' field to the mesh (pial_lh), achieves that:

![Screen Shot 2020-05-14 at 3 45 54 PM](https://user-images.githubusercontent.com/3984917/81942274-36161180-95fa-11ea-8cc1-757c29e00de7.png)

In other words, either previously our anatomical data carried an 'anatomy' field (doubt it), or this was added somewhere along the process in ft_sourceplot (but 'anatomical' isn't touched at all apart from ft_preamble).

**Comment by robertoostenveld** [comment]
Looking at other test scripts with SubjectUCI29, I notice `failed_tutorial_ecog_human_anatomy.m`

```
mac036> git log failed_tutorial_ecog_human_anatomy.m 
commit 373c08ae0f714bbd704eafd08578708f6162fecc
Author: Robert Oostenveld <r.oostenveld@gmail.com>
Date:   Wed Jul 3 13:00:02 2019 +0200

    don't use ~ for ignored output variables, see #831

commit 4d851e2d9aa05f03d77b8e0993d9fb22d18e8696
Author: Robert Oostenveld <r.oostenveld@gmail.com>
Date:   Thu Mar 28 18:33:55 2019 +0100

    renamed TEST into DEPENDENCY and cleaned up the header for a large number of test scripts (but not all)

commit 435056701758f1128524d71ab5243d835cc78ad2
Author: Nikolaas N. Oosterhof <n.n.oosterhof@googlemail.com>
Date:   Tue Jan 31 10:28:58 2017 +0100

    FIX: rename the function names of test/failed*.m to match their filename

commit 7c2e2b04527e65f030a3ab94de36cc5d2803a1d3
Author: Robert Oostenveld <r.oostenveld@gmail.com>
Date:   Tue Jan 31 08:45:21 2017 +0100

    renamed failing test scripts to "failed_xxx", see http://bugzilla.fieldtriptoolbox.org/show_bug.cgi?id=3066#c16
```

Apparently some stuff is already failing for a long time.

**Comment by robertoostenveld** [comment]
If I comment out the pos2transform on line 1151, then it fails (conceptually) later on: on line 1160 it then reads the template cortical surface (both hemispheres) rather than using the supplied surface.

**Comment by StolkArjen** [comment]
A painful reminder of how long I haven't been able to analyze data...

**Comment by StolkArjen** [comment]
I think that anatomical.anatomy field (checked for in the hasanatomical statement) is key to the solution here.

**Comment by schoffelen** [comment]
OK, the rationale behind the cfg.method = 'surface' for ft_sourceplot is:
1) the data is already on a triangulated mesh, in which case the interpolate2surf should be bypassed, since the triangulated mesh will be the surface that is to be plotted
2) the data is not on a triangulated mesh, in which case the data should be interpolated onto the surface that is to be plotted. Typically, the input 'functional' contains the funparameter and the anaparameter (where the latter is always 'anatomy'), and the surface is taken from the cfg.surffile

Arjen's idiosyncratic use case has been added later, where some bandaids seem to have been added to keep everything together. Clearly, Arjen's expectation is that an interpolate2surf can be done on-the-fly, and that the third input argument (the 'anatomical') can be used for that. The addition of the 'anatomical' has probably been inspired by the cfg.method = 'cloud' and perhaps also the plotting of ECoG grid data with cfg.method = 'surface'

I think that the logic from 341 and beyond is incomplete, because it only handles the 'cloud' method explicitly, and not the 'surface' case with hasanatomical.

**Comment by schoffelen** [comment]
It could be that if an 'anatomy' field is added to the pial surface things run through, but this is clearly not desired, because the 'anatomical' that you want to use for plotting is in this case per construction a cortically constrained mesh, which does not have an anatomy field.

**Comment by StolkArjen** [comment]
Of course, adding the anatomy field to 'anatomical' wasn't intended as a fix but to provide a clue to the cause or onset of this bug.

Don't pin me down on this, but I think the 'anatomical' was already there when we started extending fieldtrip to support human iEEG data. And rightfully so, because the ability to plot iEEG electrode data on a processed mesh (e.g., smoothed cortex, extracted hippocampus, etc), rather than on one loaded from file, has many advantages. 

The cloud plotting functionality we added later and, in fact, avoids the interpolation step around line 350. This functionality therefore has not much to do with this bug, I think.

Going forward, we should think of what we prefer ft_sourceplot and its users to do? Either load surface meshes from file, requiring also the functionality to read them from a mat file for processed meshes (perhaps that already exists), or have 'hasanatomy' yield true when 'anatomy' is a surface mesh, which reenables previous behavior of ft_sourceplot.


Answering some questions directly:
  
> Clearly, Arjen's expectation is that an interpolate2surf can be done on-the-fly, and that the third input argument (the 'anatomical') can be used for that.
Yes

>The addition of the 'anatomical' has probably been inspired by the cfg.method = 'cloud'
No

**Comment by schoffelen** [comment]
OK, well then it seems that anatomical as input was already there before the implementation of cfg.method = 'cloud', and before the on-the-fly interpolation of ECoG data. Yet, the expectation of the code has been so far that hasanatomical only runs true, if the anatomical is a genuine 3D volume with the numeric data in the 'anatomy' field. The use of a mesh for anatomical is therefore not working properly, because the code chunk that is doing the interpolation is bypassed. Removing the requirement for hasanatomical to be only true if isfield(anatomical, 'anatomy') should fix this. I guess it's easiest if I just do this, and run some tests before pushing it upstream.

**Comment by schoffelen** [comment]
Ter lering ende vermaak voor Arjen, I am currently running

ft_test('run','dependency','ft_sourceplot');

**Comment by StolkArjen** [comment]
Correct, and thanks. Alternatively, in case removing "if anatomy" breaks functionality, we could consider supplementing it with a "|| isfield(anatomical, 'tri')" statement.

**Comment by schoffelen** [comment]
| matlabversion | fieldtripversion | branch |    arch |  hostname |   user | passed | runtime |                      functionname |         mem | walltime |
|---------------|------------------|--------|---------|-----------|--------|--------|---------|-----------------------------------|-------------|----------|
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |       1 |                      test_bug1527 |  2147483648 |      600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |       9 |                      test_bug1530 |  2147483648 |      600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |      12 |                      test_bug2511 |  8589934592 |     1200 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |       1 |                      test_bug2560 |  2147483648 |      900 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |      12 |                      test_bug2769 |  2147483648 |      600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |      35 |                      test_bug3235 |  5368709120 |     1200 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |      19 |                      test_bug3275 |  2147483648 |      600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |     138 |       test_ft_read_and_plot_atlas |  3221225472 |      900 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |     431 |         test_ft_sourceinterpolate |  5368709120 |      900 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |     144 |                    test_issue1198 |  2147483648 |      600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |       4 |                    test_issue1385 |  4294967296 |     1200 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |      14 |                     test_pull1138 |  2147483648 |      600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |     769 |          test_tutorial_beamformer |  8589934592 |    12600 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |      92 |  test_tutorial_beamformer20120321 | 10737418240 |     9000 |
|         2018b |          64ae8ef | master | glnxa64 | dccn-c024 | jansch |   true |     185 | test_tutorial_beamformingextended |  5368709120 |     1800 |
